pkg/phlaredb/symdb/block_reader.go

```diff
@@ -183,15 +184,28 @@ func (r *Reader) Partition(ctx context.Context, partition uint64) (PartitionRead
 	return r.partition(ctx, partition)
 }
 
-func (r *Reader) partition(ctx context.Context, partition uint64) (*partition, error) {
+func (r *Reader) partition(ctx context.Context, partition uint64) (PartitionReader, error) {
 	p, ok := r.partitionsMap[partition]
 	if !ok {
 		return nil, ErrPartitionNotFound
 	}
-	if err := p.init(ctx); err != nil {
-		return nil, err
+	if !r.partitionsLoaded {
+		if err := p.init(ctx); err != nil {
+			return nil, err
+		}
+		return p, nil
 	}
-	return p, nil
+	return &loadedPartition{p}, nil
+}
+
+// loadedPartition is a wrapper around partition that does not
+// release the resources on Release.
+// This is used in combination with `Load` at compaction time where all partition are loaded in memory once and not released.
+type loadedPartition struct {
+	*partition
+}
+
+func (l *loadedPartition) Release() {
```

Comment:
Probably, we could avoid `*loadedPartition` as we have access to `Reader` members from the `*partition` methods and can check if it was loaded or not, like:

```go
func (p *partition) init(ctx context.Context) (err error) {
	if !p.reader.loaded {
		err = p.tx().fetch(ctx)
	}
	return err
}

func (p *partition) Release() {
	if !p.reader.loaded {
		p.tx().release()
	}
}

```

Reply:
Yeah that sounds good.

Reply:
well now I remember I did it this way because I could not get a race solved.

Reply:
Hm, which race?